File: Cpp code/2d_transformation.cpp
#include <graphics.h>
#include <math.h>
#include <conio.h>
#include <iostream>

using namespace std;

void DrawPolygon(int x[], int y[], int n) {
    for (int i = 0; i < n; i++) {
        line(x[i], y[i], x[(i + 1) % n], y[(i + 1) % n]);
    }
}

int main() {
    int gd = DETECT, gm;
    int n = 4;
    int x[4], y[4];
    int nx[4], ny[4];
    int choice;

    for (int i = 0; i < n; i++) {
        cout << "Enter point (x" << i + 1 << ", y" << i + 1 << "): ";
        cin >> x[i] >> y[i];
    }

    cout << "\n--- Enter 2D Transformation menu: ---" << endl;
    cout << "1. Translation" << endl;
    cout << "2. Rotation" << endl;
    cout << "3. Scaling" << endl;
    cout << "4. Reflection" << endl;
    cout << "5. Shearing" << endl;
    cout << "Enter Choice(1-5): ";
    cin >> choice;

    initgraph(&gd, &gm, "");

    setcolor(RED);
    DrawPolygon(x, y, n);

    setcolor(WHITE);

    if (choice == 1) {
        int tx, ty;
        cout << "Enter Translation (tx, ty): ";
        cin >> tx >> ty;

        for (int i = 0; i < n; i++) {
            nx[i] = x[i] + tx;
            ny[i] = y[i] + ty;
        }
        DrawPolygon(nx, ny, n);
    }
    else if (choice == 2) {
        float angle, rad;
        cout << "Enter Rotation (angle): ";
        cin >> angle;

        rad = angle * 3.1416 / 180.0;

        for (int i = 0; i < n; i++) {
            nx[i] = x[i] * cos(rad) - y[i] * sin(rad);
            ny[i] = x[i] * sin(rad) + y[i] * cos(rad);
        }
        DrawPolygon(nx, ny, n);
    }
    else if (choice == 3) {
        float sx, sy;
        cout << "Enter Scaling (sx, sy): ";
        cin >> sx >> sy;

        for (int i = 0; i < n; i++) {
            nx[i] = x[i] * sx;
            ny[i] = y[i] * sy;
        }
        DrawPolygon(nx, ny, n);
    }
    else if (choice == 4) {
        for (int i = 0; i < n; i++) {
            nx[i] = x[i];
            ny[i] = -y[i] + 300;
        }
        DrawPolygon(nx, ny, n);
    }
    else if (choice == 5) {
        float shx;
        cout << "Enter Shearing (shx): ";
        cin >> shx;

        for (int i = 0; i < n; i++) {
            nx[i] = x[i] + shx * y[i];
            ny[i] = y[i];
        }
        DrawPolygon(nx, ny, n);
    }

    getch();
    closegraph();
    return 0;
}

File: Python code/2d_transformation.py
import math
import matplotlib.pyplot as plt

print("Enter 4 points for the rectangle (x y):")
points = []
for i in range(4):
    x, y = map(float, input(f"Point {i+1}: ").split())
    points.append((x, y))

x_orig = [p[0] for p in points] + [points[0][0]]
y_orig = [p[1] for p in points] + [points[0][1]]

print("\n--- 2D Transformation Menu ---")
print("1. Translation")
print("2. Rotation")
print("3. Scaling")
print("4. Reflection (X-axis)")
print("5. Shearing (X-axis)")
choice = int(input("Choice: "))

fig, ax = plt.subplots()
ax.plot(x_orig, y_orig, "b--", label="Original")

if choice == 1:
    tx = float(input("Enter tx: "))
    ty = float(input("Enter ty: "))
    ax.plot(
        [x + tx for x in x_orig],
        [y + ty for y in y_orig],
        "r-",
        label="Translated",
    )
elif choice == 2:
    angle = float(input("Enter angle (in degrees): "))
    rad = math.radians(angle)
    rx = [x * math.cos(rad) - y * math.sin(rad) for x, y in zip(x_orig, y_orig)]
    ry = [x * math.sin(rad) + y * math.cos(rad) for x, y in zip(x_orig, y_orig)]
    ax.plot(rx, ry, "r-", label="Rotated")
elif choice == 3:
    sx = float(input("Enter sx: "))
    sy = float(input("Enter sy: "))
    ax.plot(
        [x * sx for x in x_orig], [y * sy for y in y_orig], "r-", label="Scaled"
    )
elif choice == 4:
    ax.plot(x_orig, [-y for y in y_orig], "r-", label="Reflected")
    ax.axhline(0, color="black", linewidth=1)
elif choice == 5:
    shx = float(input("Enter shx: "))
    ax.plot(
        [x + shx * y for x, y in zip(x_orig, y_orig)],
        y_orig,
        "r-",
        label="Sheared",
    )

ax.set_aspect("equal")
plt.grid(True)
plt.legend()
plt.show()
